### ISSUE TITLE
Enhancement: prevent wave propagation for the given edges

### ISSUE THREAD

**Issue by vvoovv** [prochitecture/bpypolyskel#9]
Sometimes it's desirable to prevent wave propagation for some edges.

Below are a couple of examples.

![image](https://user-images.githubusercontent.com/613295/103299780-5d491400-4a0e-11eb-931b-50563999af22.png)

![image](https://user-images.githubusercontent.com/613295/103299715-35f24700-4a0e-11eb-8675-a2132545f603.png)

The red color is used mark the edges with no wave propagation.

**Comment by polarkernel** [comment]
> Sometimes it's desirable to prevent wave propagation for some edges.
Do you think it would be possible to include the feature #9 to bpypolyskel?

No, I don't think so, I see no way.

**Comment by vvoovv** [comment]
If there are edges that don't emit waves, edge events for those edges can be omitted. However the other events must be calculated for those edges.

If there is straight skeleton edge with a loose end (like in the example below), then its intersection with an edge of the original polygon must be found.

![image](https://user-images.githubusercontent.com/613295/103299780-5d491400-4a0e-11eb-931b-50563999af22.png)

**Comment by polarkernel** [comment]
> If there are edges that don't emit waves, edge events for those edges can be omitted. However the other events must be calculated for those edges.

Unfortunately it is not that simple. Let me explain using an example. In the following rectangle, I removed the edge events of the edge at the right, while the vertexex remained valid (left image, click on images to magnify). When the edge event 0 is processed (middle image), it creates a new edge event 4 (blue dot), which finally, when processed, leads to the complete skeleton (right). The red lines show the list of active vertices (LAV).

<img src="https://user-images.githubusercontent.com/48180714/104123768-5941be00-534d-11eb-94e1-9fcb25b64331.png" width="260"> <img src="https://user-images.githubusercontent.com/48180714/104123810-868e6c00-534d-11eb-9e32-82f11c663581.png" width="260"> <img src="https://user-images.githubusercontent.com/48180714/104123875-d2d9ac00-534d-11eb-89bc-9dd1f1f76193.png" width="260">

If the vertexes of the edge not to be moved get invalidated (shown as red cross instead of dot), we get the desired result for this simple case.  Again, the edge event 0 creates a new edge event 4, but this time it is invalid (blue cross):

<img src="https://user-images.githubusercontent.com/48180714/104123768-5941be00-534d-11eb-94e1-9fcb25b64331.png" width="260"> <img src="https://user-images.githubusercontent.com/48180714/104124111-32848700-534f-11eb-8fa9-72b16996b8ca.png" width="260"> <img src="https://user-images.githubusercontent.com/48180714/104124117-36b0a480-534f-11eb-8086-4795bd135ece.png" width="260"> 

But the LAV contains now invalid vertices, which should not be the case. Removing them lets the LAV collapse to a line. It worked for this simple case, but here is an example, where the process crashes:

<img src="https://user-images.githubusercontent.com/48180714/104124428-2ac5e200-5351-11eb-98d6-185f5ed523e3.png" width="260"> <img src="https://user-images.githubusercontent.com/48180714/104124432-2f8a9600-5351-11eb-8f2e-0163052cf12a.png" width="450">

I assume that in this case the reason was a split event that used one of the original edges as opposite edge. These edges (not the same as the edges we enter to skeletonize) don't get actualized for validity of vertexes.

**Comment by polarkernel** [comment]
> Sometimes it's desirable to prevent wave propagation for some edges.

Could we eventually solve this using a postprocessing at the very end of _polygonize()_. At least for the simple cases it should easily be possible to identify the edge, as their indices ar the first two of the face, while the third is the node to be changed. Moving this node to the center of the edge while keeping its height should bring the desired result, a vertical face:

<img src="https://user-images.githubusercontent.com/48180714/104124858-76798b00-5353-11eb-9b9d-4bdbbf5003a5.png" width="300"> <img src="https://user-images.githubusercontent.com/48180714/104124864-79747b80-5353-11eb-97e2-9010c48ef818.png" width="300">

I do not yet understand your second example, the image is too blurry for my eyes. May I get the corresponding footprint?

**Comment by vvoovv** [comment]
I created a simplified version (no spikes and perfect alignment) of the desired result. I'll add the footprint later.

![image](https://user-images.githubusercontent.com/613295/104129045-ea2f8e00-537b-11eb-9f64-b95974959e22.png)

Below is "normal" version for the comparison:
![image](https://user-images.githubusercontent.com/613295/104444088-58f12f00-55a8-11eb-85a9-9ac26f92d4c2.png)

**Comment by vvoovv** [comment]
> May I get the corresponding footprint?

It's actually already available in the automated tests:
[tests/test_3803883_salamanca_avenida_filiberto_villalobos.py](https://github.com/prochitecture/bpypolyskel/blob/master/tests/test_3803883_salamanca_avenida_filiberto_villalobos.py)

**Comment by vvoovv** [comment]
> When the edge event 0 is processed (middle image), it creates a new edge event 4 (blue dot) ...
> 
> <img alt="" width="260" src="https://user-images.githubusercontent.com/48180714/104123875-d2d9ac00-534d-11eb-89bc-9dd1f1f76193.png">

Maybe I don't understand something. I think the edge event 4 can't be created, since the rightmost edge doesn't emit a wavefront. And the triangle with red edges can't happen in this case since the the rightmost edge isn't available to form the triangle.

**Comment by vvoovv** [comment]
> Could we eventually solve this using a postprocessing at the very end of _polygonize()_.

It was my original idea. Yes, it would work for the triangles formed by the straight skeleton but not for more advanced cases like the one in Salamanca.

**Comment by polarkernel** [comment]
> Maybe I don't understand something.

The first row of images is the example that follows your proposition 
> If there are edges that don't emit waves, edge events for those edges can be omitted. <ins>However the other events must be calculated for those edges</ins>.

The algorithm does not really work with wavefronts. There are only vertexes, which are used to compute new events. In this example (first row of images) I left the vertexes belonging to the static edge as unprocessed (or _valid_ in Botffy's code), while I removed its events. Therefore, the vertexes are used to create the egde event 4.

In the second example (second row of images) I marked these vertexes as processed (invalid). Then the egde event 4 gets created, but is invalid because its vertexes are invalid. But in more complicated contours, also for Salamanca, this leads to crashes.

**Comment by vvoovv** [comment]
Ok, I see. That's because I haven't still studied the original algorithm.


I have still a note to the image below. The edge event happens in the intersection point of bisectors. However the bisector for the upper red vertex of the triangle with red edges isn't depicted correctly. Is my note correct?

<img alt="" width="260" src="https://user-images.githubusercontent.com/48180714/104123875-d2d9ac00-534d-11eb-89bc-9dd1f1f76193.png">

**Comment by polarkernel** [comment]
> I have still a note to the image below.

You are right, my images are not that clear. I use them during the debugging and the left one just shows me, where the event to be processed came from. Let me try to explain the process in detail, so that you hopefully get a clearer picture.

A _Vertex_ consists of the position of a vertice, the previous and next edge, the bisector between these edges and links to the previous and next _Vertex_. Initially there have been four vertexes, one for every vertice of the rectangle. The intersection of the bisectors of the two vertexes on the left created the event 0 during initialization. Processing the event 0 goes through the following steps:

1. A new vertex is created at the position of the event.
2. The two vertexes of event 0 are removed from the circular list LAV, while the new vertex is inserted. The LAV is now the red line in the image.
3. The two vertexes that created event 0 are marked as invalid (processed).
4. Because the new vertex is not reflex, a new edge event 4 is created at the intersection of the bisector of the new vertex and the bisector of the previous vertex.
5. The same happens for the bisectors of the new and the next vertex, but this event 5 is not processed, because the remaining vertexes get invalidated when event 4 is processed.

What I depicted on the left part of the image was the LAV, before event 4 was processed, and the bisectors (dotted lines) that constructed the event 4 in step 4. The LAV is not used to for bisectors, it just remembers the ordering (previous and next) of the  vertexes. All processing steps use always the original edges, which makes things sometimes so complicated, mainly for parallel edges.

**Comment by vvoovv** [comment]
> 4. Because the new vertex is not reflex, a new edge event 4 is created at the intersection of the bisector of the new vertex and the bisector of the previous vertex.

Just to make things even more clear. The _previous vertex_ in the sentence above is the 90 degrees vertex of the original rectangle. Right?

**Comment by polarkernel** [comment]
> Right?

Exactly! To find the previous vertex one follows the LAV (red) from the new vertex in clockwise direction.

**Comment by polarkernel** [comment]
I can't find a solution for this request. I tried a lot of ideas, but most of them failed already on paper. The skeletonize algorithm does not provide any handle to reach this goal. I even tried to move the vertices of the static edge to "infinity".

I think the only way would be to use another type of skeletonize algorithm. It seems that for instance a weighted straight skeleton could do this job by setting the weight of the static edge to zero. I read once also that a straight skeleton based on a motorscycle graph could do that, but I was no more able to retrieve this text.

**Comment by vvoovv** [comment]
> It seems that for instance a weighted straight skeleton could do this job by setting the weight of the static edge to zero.

Or in the terms of the wavefront, setting the wavefront speed to _V_ for all "normal" edges and to zero for the static edges.

**Comment by vvoovv** [comment]
By the way, found a very detailed [explanation](https://neerc.ifmo.ru/wiki/index.php?title=Straight_skeleton#.D0.90.D0.BB.D0.B3.D0.BE.D1.80.D0.B8.D1.82.D0.BC_.D1.81_.D0.B8.D0.B7.D0.BF.D0.BE.D0.BB.D1.8C.D0.B7.D0.BE.D0.B2.D0.B0.D0.BD.D0.B8.D0.B5.D0.BC_SLAV) of the SLAV algorithm in Russian.

**Comment by polarkernel** [comment]
> Or in the terms of the wavefront, setting the wavefront speed to V for all "normal" edges and to zero for the static edges.

Exactly. In literature about weighted straight skeletons, the terms _speed_, _weight_ and _angle_ (of the roof's slope) are used interchangebly. However, events do no more lie on bisector intersections and that's why our algorithm may not be adapted for this case.

**Comment by polarkernel** [comment]
> By the way, found a very detailed explanation of the SLAV algorithm in Russian.

Interesting. These are excerpts of papers and presentations mainly by Felkel and Obdržálek, Tom Kelly, Stefan Huber and Martin Held, using their original figures. I found nothing new to me, I already have all these papers.

One amusing sentence, not from one of these papers:  Ещё примеры не для слабонервных.  ;-)

**Comment by vvoovv** [comment]
> I read once also that a straight skeleton based on a motorscycle graph could do that, but I was no more able to retrieve this text.

I assume it was the paper by Stefan Huber. Here is the list of his papers with the download links:
https://www.sthu.org/research/publications/
There are quite a few of papers in the list.

**Comment by polarkernel** [comment]
> I assume it was the paper by Stefan Huber.

Maybe. [Martin Held](https://www.cosy.sbg.ac.at/~held/held.html), [Stefan Huber](https://www.sthu.org/) and [Peter Palfrader](www.palfrader.org) form a very active group researching around straight skeletons at the _Department of Computer Science of the University of Salzburg_ in Austria. Martin Held is Director of the _Computational Geometry and Applications Lab_ there. Stefan Huber is professor in this departement, while Peter Palfrader is postdoc research fellow, again in the same departement.

They do an excellent and very active research on straight skeletons and have written tons of papers around this subject. There main goal seems to be to crack the best big O values. They also provide two open source implementations, [MONOS ](https://github.com/cgalab/monos) and [SURFER2](https://github.com/cgalab/surfer2), based on [CGAL](https://www.cgal.org/). It is all written in C++ and it is absolutely impossible to incorporate CGAL to Blender.

Unfortunately, my skills to understand and transform their mathematical notations to code are quite limited.

**Comment by vvoovv** [comment]
I am more than happy with the current state of bpypolyskel!

**Comment by costika1234** [comment]
Guys, this library is an absolute jewel. I can't thank you enough for all the hard work you put into creating a developer friendly tool that handles so many roof-related edge cases!

**Comment by vvoovv** [comment]
Hi @costika1234 

Thank you for your feedback.

May I know how you are using _bpypolyskel_?

**Comment by costika1234** [comment]
I used this library to build a SketchUp extension that generates a roof from a 2D plan. An existing SketchUp plugin (from 1001bit Tools) was unable to generate the correct geometry for a biscuit-like shape, but it looks like I can achieve this task via your amazing library. 

That said, the correct geometry seems to depend on the position of the 2D polygon within the SketchUp canvas. Perhaps some precision errors are creeping in, but I can just translate the polygon a bit to arrive at the desired output. I will play a bit more with this example and will let you know if I find any library-related problems.

PS. Take a look at the PR I posted earlier today! 😉
